src/v/utils/chunked_kv_cache.h

```diff
@@ -0,0 +1,195 @@
+
+    /**
+     * Gets the key's corresponding value from the cache.
+     *
+     * Returns std::nullopt if the key doesn't have a value in the cache.
+     */
+    ss::optimized_optional<ss::shared_ptr<Value>> get_value(const Key& key);
+
+    using cache_stat = struct cache_t::stat;
+    /**
+     * Cache statistics.
+     */
+    struct stat : public cache_stat {
+        /// Current size of the cache index.
+        size_t index_size;
+    };
+
+    /**
+     * Returns the current cache statistics.
+     */
+    [[nodiscard]] stat stat() const noexcept;
+
+private:
+    using ghost_hook_t = boost::intrusive::list_member_hook<
+      boost::intrusive::link_mode<boost::intrusive::safe_link>>;
+
+    struct cached_value {
+        Key key;
+        ss::shared_ptr<Value> value;
+        s3_fifo::cache_hook hook;
+        ghost_hook_t ghost_hook;
+    };
+
+    using entry_t = std::unique_ptr<cached_value>;
+    using ghost_fifo_t = boost::intrusive::list<
+      cached_value,
+      boost::intrusive::
+        member_hook<cached_value, ghost_hook_t, &cached_value::ghost_hook>>;
+
+    chunked_hash_map<Key, entry_t, Hash, EqualTo> _map;
+    cache_t _cache;
+    ghost_fifo_t _ghost_fifo;
+
+    void gc_ghost_fifo();
+};
+
+template<typename Key, typename Value, typename Hash, typename EqualTo>
+struct chunked_kv_cache<Key, Value, Hash, EqualTo>::evict {
+    chunked_kv_cache& kv_c;
+
+    bool operator()(cached_value& e) noexcept {
+        e.value = nullptr;
+        kv_c._ghost_fifo.push_back(e);
+        return true;
+    }
+};
+
+template<typename Key, typename Value, typename Hash, typename EqualTo>
+bool chunked_kv_cache<Key, Value, Hash, EqualTo>::try_insert(
+  const Key& key, ss::shared_ptr<Value> val) {
+    gc_ghost_fifo();
+
+    auto e_it = _map.find(key);
+    if (e_it == _map.end()) {
+        auto [e_it, succ] = _map.try_emplace(
+          key, std::make_unique<cached_value>(key, std::move(val)));
+        if (!succ) {
+            return false;
+        }
+
+        _cache.insert(*e_it->second);
+        return true;
+    }
+
+    auto& entry = *e_it->second;
+    if (entry.hook.evicted()) {
+        entry.value = std::move(val);
+        _ghost_fifo.erase(_ghost_fifo.iterator_to(entry));
+        _cache.insert(entry);
+        return true;
+    }
+
+    return false;
+}
+
+template<typename Key, typename Value, typename Hash, typename EqualTo>
+ss::optimized_optional<ss::shared_ptr<Value>>
+chunked_kv_cache<Key, Value, Hash, EqualTo>::get_value(const Key& key) {
+    gc_ghost_fifo();
+
+    auto e_it = _map.find(key);
+    if (e_it == _map.end()) {
+        return std::nullopt;
+    }
+
+    auto& entry = *e_it->second;
+    if (entry.hook.evicted()) {
+        return std::nullopt;
+    }
+
+    entry.hook.touch();
+    return entry.value;
+}
+
+template<typename Key, typename Value, typename Hash, typename EqualTo>
+void chunked_kv_cache<Key, Value, Hash, EqualTo>::gc_ghost_fifo() {
```

Comment:
yeh when to GC ghost entries is an interesting question, and is largely left as an exercise to the user because it's intention is to be domain specific (e.g. specific to different indexing strategies or use cases). i don't have any opinion about this one, it seems ok. just something to note that maybe would change over time for example if had some degraded mode where the linear scan in it was undesirable. in which case perhaps the costs of GC could be amortized more smoothly across operations, or be done by a background fiber.

but overall i guess the assumption is that the ghost queue entries themselves are not the big thing that memory consumers that GC addresses, that is the actual `value`s that are removed on insert when the cache is full?

Reply:
> just something to note that maybe would change over time for example if had some degraded mode where the linear scan in it was undesirable

Might be misunderstanding what you said, but it should never actually scan over all ghost fifo entries. In cases where its called and no entries have expired then it'll just check the first entry for expiration then return. In cases where its called and N entries have expired then it'll iterate over those N entries to remove them from the index plus an additional entry then return. This is since the fifo orders the entries from oldest to newest. 

I guess in the worst case where all the entries in the ghost fifo have expired you may not want to immediately remove all the entries in the next GC call and instead split them up among the next N if that's what you meant.

Reply:
> but overall i guess the assumption is that the ghost queue entries themselves are not the big thing that memory consumers that GC addresses, that is the actual values that are removed on insert when the cache is full?

Yeah, its the main assumption that the values end up consuming far more memory than the index and entries. 

A bit unrelated, but I've considered requiring that users provide a `size` function for values in the cache. This could allow the cache's configuration to be in maximum bytes instead of maximum entries. It would also allow us to ensure the cache stays under a certain size even if the assumption that values are larger than the index/entries doesn't hold. Lastly we could potentially hook the cache up to seastar's memory reclaim.

However, it may be complicated for some value types to actually determine their memory usage. This ended up being the case for schema types in datalake.

Reply:
>I guess in the worst case where all the entries in the ghost fifo have expired you may not want to immediately remove all the entries in the next GC call and instead split them up among the next N calls if that's what you meant.

Yes, that's what I meant--reactor stall risk. But I agree it's not a big concern. It'd be hard to hit frequently, and if we did encounter an issue, we could probably adjust some things to address it.

Reply:
>maximum bytes instead of maximum entries

Ahh yeh I had an email exchange with the inventor of s3fifo with a similar question. They basically said that the assumption was that there was some uniformity in the entry sizes, or at least, the the maximum size of entries wasn't too large. They also mentioned that supporting heterogeneity was something they were considering for future research.